In [3]:
import pandas as pd
import numpy as np
news_data = pd.read_csv('/content/newss.csv',index_col=False, dtype='unicode',low_memory = False)

# 1. Introduction

As per the current scenario of social media and every kind of internet-related things people are totally depended on that sometimes we don’t know that every news and articles are not a real thing which happened

In the world but we were believed in that social media is the largest user base platform which consists the news that sometimes real or fake this system identifies that every kind of fake and real news with a powerful platform of data science and also uses a large amount of dataset which consists lots of news related data by analytical platforms.

# What is FAKE NEWS?

A type of yellow journalism, fake news encapsulates pieces of news that may be hoaxes and is generally spread through social media and other online media. This is often done to further or impose certain ideas and is often achieved with political agendas. Such news items may contain false and/or exaggerated claims, and may end up being virtualized by algorithms, and users may end up in a filter bubble.**

# What is TfidfVectorizer?

· TF (Term Frequency): The number of times a word appears in a document is its Term Frequency. A higher value means a term appears more often than others, and so, the document is a good match when the term is part of the search terms.

· IDF (Inverse Document Frequency): Words that occur many times a document, but also occur many times in many others, maybe irrelevant. IDF is a measure of how significant a term is in the entire corpus.

The TfidfVectorizer converts a collection of raw documents into a matrix of TF-IDF features.

In [ ]:
import pandas as pd
import numpy as np
news_data = pd.read_csv('/content/newss.csv',index_col=False, dtype='unicode',low_memory = False)

# What is PassiveAggressiveClassifier?

The passive-aggressive algorithms are a family of algorithms for large-scale learning. They are similar to the Perceptron in that they do not require a learning rate. However, contrary to the Perceptron, they include a regularization parameter C.

In [4]:
news_data.head()

Unnamed: 0                                              title  \
0       8476                       You Can Smell Hillary’s Fear   
1      10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2       3608        Kerry to go to Paris in gesture of sympathy   
3      10142  Bernie supporters on Twitter erupt in anger ag...   
4        875   The Battle of New York: Why This Primary Matters   

                                                text label Unnamed: 4  \
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE        NaN   
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE        NaN   
2  U.S. Secretary of State John F. Kerry said Mon...  REAL        NaN   
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE        NaN   
4  It's primary day in New York and front-runners...  REAL        NaN   

  Unnamed: 5 Unnamed: 6 Unnamed: 7 Unnamed: 8 Unnamed: 9  ... Unnamed: 131  \
0        NaN        NaN        NaN        NaN        NaN  ...          NaN   
1        NaN        NaN        NaN        NaN        NaN  ...          NaN   
2        NaN        NaN        NaN        NaN        NaN  ...          NaN   
3        NaN        NaN        NaN        NaN        NaN  ...          NaN   
4        NaN        NaN        NaN        NaN        NaN  ...          NaN   

  Unnamed: 132 Unnamed: 133 Unnamed: 134 Unnamed: 135 Unnamed: 136  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          NaN          NaN   
2          NaN          NaN          NaN          NaN          NaN   
3          NaN          NaN          NaN          NaN          NaN   
4          NaN          NaN          NaN          NaN          NaN   

  Unnamed: 137 Unnamed: 138 Unnamed: 139 Unnamed: 140  
0          NaN          NaN          NaN          NaN  
1          NaN          NaN          NaN          NaN  
2          NaN          NaN          NaN          NaN  
3          NaN          NaN          NaN          NaN  
4          NaN          NaN          NaN          NaN  

[5 rows x 141 columns]

In [10]:
df = news_data.drop(news_data.loc[:,'Unnamed: 4':'Unnamed: 140'].columns, axis=1)

**Step-1: Download First Dataset of news to work with real-time data**

The dataset we’ll use for this python project- we’ll call it news.csv. This dataset has a shape of 7795×4. The first column identifies the news, the second and third are the title and text, and the fourth column has labels denoting whether the news is REAL or FAKE

**Step-2: Make Necessary Imports**

In [11]:
df.shape

(7795, 4)

# Step-4: And Get labels from DataFrame

In [12]:
labels = df.label
labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

# Step-5: Split the dataset into training and testing sets.

In [13]:
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test = train_test_split(df['text'],labels, test_size=0.2, random_state=7)

# Step-6: Let’s initialize a TfidfVectorizer with stop words from the English language and a maximum document frequency of 0.7 (terms with a higher document frequency will be discarded). Stop words are the most common words in a language that is to be filtered out before processing the natural language data. And a TfidfVectorizer turns a collection of raw documents into a matrix of TF-IDF features.

# # Initialize a TfidfVectorizer

tfidf_vectorizer=TfidfVectorizer(stop_words=’english’, max_df=0.7)
# Fit and transform train set, transform test set
tfidf_train=tfidf_vectorizer.fit_transform(x_train)
tfidf_test=tfidf_vectorizer.transform(x_test)

#Initialize a PassiveAggressiveClassifier
pac=PassiveAggressiveClassifier(max_iter=50)
pac.fit(tfidf_train,y_train)
#DataPredict on the test set and calculate accuracy
y_pred=pac.predict(tfidf_test)
score=accuracy_score(y_test,y_pred)
print(f’Accuracy: {round(score*100,2)}%’)

Now, fit and transform the vectorizer on the train set, and transform the vectorizer on the test set.

In [23]:
import pandas as pd

# Assuming X_train is a pandas DataFrame
X_train.fillna('', inplace=True)
X_test.fillna('', inplace=True)


In [24]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Assuming X_train is preprocessed (NaN values filled or removed)
tfidf_vectorizer = TfidfVectorizer()
tfidf_train = tfidf_vectorizer.fit_transform(X_train)
tfidf_test = tfidf_vectorizer.transform(X_test)


In [25]:
tfidf_vectorizer=TfidfVectorizer(stop_words='english', max_df=0.7)

In [28]:
from sklearn.linear_model import PassiveAggressiveClassifier

# Create a PassiveAggressiveClassifier instance
pac = PassiveAggressiveClassifier(max_iter=50)

y_train.fillna('', inplace=True)
y_test.fillna('', inplace=True)

# Fit the classifier on the TF-IDF transformed training data and labels
pac.fit(tfidf_train, y_train)


PassiveAggressiveClassifier(max_iter=50)

In [29]:
y_pred=pac.predict(tfidf_test)


from sklearn.metrics import accuracy_score

# Calculate the accuracy score
score = accuracy_score(y_test, y_pred)

# Print the accuracy score
print(f"Accuracy Score: {score}")



Accuracy Score: 0.7601026298909558


# Step-7: Now, we will initialize the PassiveAggressiveClassifier This is. We’ll fit this on tfidf_train and y_train.

# Step-8: Now after the Accuracy computation we have to build a confusion matrix.

In [14]:

from sklearn.metrics import accuracy_score, confusion_matrix

In [31]:
confusion_matrix(y_test,y_pred,labels=['FAKE','REAL'])

array([[577,  38],
       [ 42, 603]])

**Conclusion**

I hereby declared that my system detecting Fake and real news from a given dataset with 76.01% Accuracy Level. And also solve the issue of Yellow Journalism